File: 17.电话号码的字母组合.py
#
# @lc app=leetcode.cn id=17 lang=python3
#
# [17] 电话号码的字母组合
#

# @lc code=start
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []

        d = {
            '2':'abc', '3':'def', '4':'ghi', '5':'jkl', '6':'mno', 
            '7':'pqrs', '8':'tuv', '9':'wxyz'
        }
        size = len(digits)
        ans = []
        s = ''

        def solve(s, p):
            if p == size:
                ans.append(s)
                return
            alpha = d[digits[p]]
            for i in range(len(alpha)):
                s += alpha[i]
                solve(s, p+1)
                s = s[:-1]
        
        solve(s, 0)
        return ans
                    

# @lc code=end

File: 17.电话号码的字母组合.java
/*
 * @lc app=leetcode.cn id=17 lang=java
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start
class Solution {
    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0) {
            return new ArrayList<String>();
        }
        List<String> ans = new ArrayList<>();
        String[] tel = new String[10];
        tel[2] = "abc";
        tel[3] = "def";
        tel[4] = "ghi";
        tel[5] = "jkl";
        tel[6] = "mno";
        tel[7] = "pqrs";
        tel[8] = "tuv";
        tel[9] = "wxyz";

        getComb(ans, digits, 1, tel, "");
        return ans;
    }

    private void getComb(List<String> ans, String digits, int level, String[] tel, String s) {
        if (level == digits.length()) {
            int n = digits.charAt(level-1)-'0';
            for (char c: tel[n].toCharArray()) {
                ans.add(s + c);
            }
            return;
        }
        int n = digits.charAt(level-1)-'0';
        for (char c: tel[n].toCharArray()) {
            getComb(ans, digits, level+1, tel, s+c);
        }
    }
}
// @lc code=end

